### DISCUSSION TITLE
Data Shape for Training and Prediction

### DISCUSSION THREAD

**Discussion by nyanko-box** [awslabs/gluonts#514] [Q&A]
Hi, there.

I got confusing about the shape of dataset.
Could you please confirm my understanding?

**Assuming that I use `DeepAREstimator()` with `feat_dynamic_real`.**

## training
Figure in below shows a set of target and related variables in context and prediction ranges which I understand.

![training](https://user-images.githubusercontent.com/5070374/70601923-931a2f80-1c36-11ea-97d5-236de2017c3a.PNG)

The target in context range [1:t0] and the covariates in context and prediction ranges [1:T] are used to calculate the Temporary predicted values.
The Temporary predicted values are compared with the targets in prediction range [t0+1:T],
and then the training progresses.

## prediction
According to the issue #468 , @mbohlkeschneider said that

> the predict method would expect that your target has some length n and the dynamic features n + prediction length.

I think that
* this `n` equals `context_length` which is in my example
* the data structure in prediction is shown as below figure.

![prediction](https://user-images.githubusercontent.com/5070374/70604583-3752a500-1c3c-11ea-90e6-1c2dd5686d92.PNG)

# Question
My understanding about data shape on training and prediction processes are correct?

Thanks in advance.

**Comment by jaheba** [comment]
Hey @nyanko-box,

during training of DeepAR "training-windows" are sliced out of the training-dataset. Note, that the training dataset need not be uniform, i.e. time-series can start end end at any point.

What is correct is that we use the last `prediction_length`-values of the target to evaluate and improve the model on.

Note, that DeepAR gets to see all past-data, not just `context_length` and makes use of it through lagged-values.

For prediction, you need to provide `prediction_length` more items in the covariate time-series (compared to target), since they are part of the input for the network.

**Comment by nyanko-box** [comment]
Hi, @jaheba 

Thank you for replying.
I have a question for training process.

> during training of DeepAR "training-windows" are sliced out of the training-dataset
...
the last prediction_length-values of the target to evaluate and improve the model on.

I think about training as:
the whole training-dataset with `dataset_length` is divided into some parts with the length of (`context_length`+`prediction_length`). With your words, the length of (`context_length`+`prediction_length`) is the length of "training-windows."
Then, the latest part is used to evaluate and improve.
Are these correct? 

---

> For prediction, you need to provide prediction_length more items in the covariate time-series (compared to target), since they are part of the input for the network.

OK. I clearly understood for prediction. Thank you.
